Add documentation for image-builder github action (#10872)
kyma-project/test-infra: .github/actions/image-builder/README.md

Reviewer: Please consider replacing the list of inputs and outputs with link to the code and moving the description to the code itself. This way we will not have update list every time we change inputs or outputs. Moreover, the detailed description will exist together with code.

Should we mention that this action requires some environment variables to be present and that it uses these variables values to set values of parameters passed to the ADO pipeline?

Author: The url must be maintained as well, will we remember about that?
I definitely prefer to create this section with some automation. As for now, when the action is used only by our reusable workflow I will make that change.

# Image Builder Action

This action is supposed to trigger image-builder service. It receives authentication tokens (**oidc** and **azure personal access token**) via inputs as well as image name.

Optionally you can set other inputs to control which image with which context should be built.

# Inputs

> ## **oidc-token**

Author: I would like to leave it as it follow the expose-jwt style.

> JWT token exposed by **expose-jwt** github action, which is JWT token signed by github, containing information about running workflow.

Author: ```suggestion
A OIDC token exposed by the **expose-jwt** GitHub action, which is a JWT token signed by GitHub, containing information about running workflow.
```

Required: **true**

## **ado-token**

Azure DevOps personal access token (PAT), which is used by this action to authenticate against Azure API.

Required: **true**

## **image-name**

Name of the image that will be build via image builder.

Required: **true**

## **context**

Optional parameter that allows you to set build context for image builder, related to the root of the repository.

Default: **'.'**

## **dockerfile**

Optional parameter that allows you to set path to the Dockerfile for the image to build.

Default: **'Dockerfile'**

## **build-args**

Optional parameter that allows you to set additional build arguments.
Each argument has format `name=value`, multiple arguments are separated by new line.

Default: **''**

## **tags**

> Optional parameter that allows you to pass tags with which image will be build. Each tag is placed in the new line.

Reviewer 2: ```suggestion
Optional parameter to pass tags with which the image is built. Each tag is placed in a new line.
```

Default: **''**

## **export-tags**

> Optional parameter that enable exporting tags provided via **tags** input and default tags to be exported as build argument.

Reviewer 2: ```suggestion
Optional parameter that enables exporting tags provided with the **tags** input and default tags to be exported as build argument.
```

> Each tag will have format **TAG_x**, where `x` is the tag name passed along with the tag.

Reviewer 2: ```suggestion
Each tag gets the format **TAG_x**, where `x` is the tag name passed along with the tag.
```

Default: **false**

## **config**

> Optional parameter which sets the config file containing information about connection to the Azure DevOps.

Reviewer 2: ```suggestion
Optional parameter that sets the config file containing information about connection to the Azure DevOps.
```

Default: **'./configs/kaniko-build-config.yaml'**

## **env-file**

> Optional parameter to provide a path to the file with environment variables that loads in the build.

Reviewer 2: ```suggestion
Optional parameter that provides the path to the file with environment variables that are loaded into the build environment.
```

> Optional parameter that allows you to provide path to the file with environment variables which loads in the build

Reviewer 2: ```suggestion
Optional parameter to provide a path to the file with environment variables that loads in the build.
```
What is the thing that does the loading - the file, or the environment variables?
If it's the variables, you need plural:

```suggestion
Optional parameter to provide a path to the file with environment variables that load in the build.
```

Author: It's more like 'that image-builder loads during the build'. We're loading variables from provided file path into azure devops environment.

Default: **''**

## **dry-run**

> Optional parameter that allows you to prevent calling Azure service

Reviewer 2: We need an article, but I don't understand the context enough to know which
Either
```suggestion
Optional parameter to prevent calling an Azure service
```
Or
```suggestion
Optional parameter to prevent calling the Azure service
```

Default: **false**

# Outputs

## **adoResult**

Result status of the Azure DevOps execution

## **images**

JSON formatted array containing all images build during process.

> # How It Work?

Reviewer 3: ```suggestion
# How It Works?
```

> Github action is using **europe-docker.pkg.dev/kyma-project/prod/image-builder** docker image to trigger pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO and kaniko-build-config from `main` branch of `test-infra` repository.

Reviewer 2: ```suggestion
The image builder action uses a **europe-docker.pkg.dev/kyma-project/prod/image-builder** Docker image to trigger a pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO and kaniko-build-config from the `main` branch of `test-infra` repository.
```

> The image builder action uses a **europe-docker.pkg.dev/kyma-project/prod/image-builder** Docker image to trigger a pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO and kaniko-build-config from the `main` branch of `test-infra` repository.

Reviewer: My understanding of this sentence is that ADO and kaniko-build-config provide parameters values which are passed to the REST API. This is not true. Could you please rephrase it to make it clear that ADO expose some REST API and we use data from kaniko-build-config to access this API.

Author: Sure

> Github action is using **europe-docker.pkg.dev/kyma-project/prod/image-builder** docker image to trigger pipeline in Azure DevOps (ADO). It's passing parameters using REST API provided by ADO and kaniko-build-config from `main` branch of `test-infra` repository.

Reviewer 3: ```suggestion
Github action is using **europe-docker.pkg.dev/kyma-project/prod/image-builder** docker image to trigger pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO and kaniko-build-config from `main` branch of `test-infra` repository.
```

> During the ADO pipeline execution this action is checking for status to be reported, if execution ended it fetches status and logs.

Reviewer 2: ```suggestion
During the ADO pipeline execution, the image builder action is checks for the status to be reported. When the execution ends, it fetches the status and logs.
```

> During the ADO pipeline execution is checking for status to be reported, if execution ended it fetches status and logs.

Reviewer 3: ```suggestion
During the ADO pipeline execution this action is checking for status to be reported, if execution ended it fetches status and logs.
```

> If execution failed, the github action fails. If the execution successes, it sets the output and success.

Reviewer 2: ```suggestion
If the execution failed, the the image builder action fails. If the execution succeeds, it sets the output and success status.
```
I'm guessing "success" is a kind of status

Author: That's correct

> If execution failed, the github action is also failing. If the execution successed, it sets the output and success.

Reviewer 3: ```suggestion
If execution failed, the github action fails. If the execution successes, it sets the output and success.
```

# Example Usage

```
- uses: ./.github/actions/image-builder
        id: build
        name: Run build in image-builder
        with:
          oidc-token: 'oidc-token'
          ado-token: ${{ secrets.ADO_PAT }}
          context: '.'
          build-args: |
            arg1=value1
            arg2=value2
          tags: |
            latest
            main
            commit=${{ .ShortSHA }}
          export-tags: true
          image-name: 'ginkgo'
          dockerfile: 'Dockerfile'

>           env-file: '.env'

Reviewer 3: ```suggestion
          env-file: 'envs'
```

# Image Builder Action

> The image builder action triggers the image-builder service. It receives authentication tokens (**oidc** and **azure personal access token**) from inputs as well as from the image name.

Reviewer 4: ```suggestion
The image-builder action triggers the image-builder service. It receives authentication tokens (**oidc** and **azure personal access token**) from inputs as well as from the image name.
```

Optionally, you can set other inputs to control which image with which context should be built.

# Inputs

> Description of each input is available [here](https://github.com/kyma-project/test-infra/blob/main/.github/actions/image-builder/action.yml#L3-L44).

Reviewer 4: ```suggestion
Descriptions of each input is available in the [`action`](https://github.com/kyma-project/test-infra/blob/main/.github/actions/image-builder/action.yml#L3-L44) file.
```

# Outputs

> Description of each output is available [here](https://github.com/kyma-project/test-infra/blob/main/.github/actions/image-builder/action.yml#L46-L52).

Reviewer 4: ```suggestion
Descriptions of each output is available in the [`action`](https://github.com/kyma-project/test-infra/blob/main/.github/actions/image-builder/action.yml#L46-L52) file.
```

# How It Works?

> The image builder action uses a **europe-docker.pkg.dev/kyma-project/prod/image-builder** Docker image to trigger a pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO. The definition of parameters passed via REST API is taken from kaniko-build-config located on the `main` branch of `test-infra` repository and github context.

Reviewer 4: ```suggestion
The image-builder action uses a **europe-docker.pkg.dev/kyma-project/prod/image-builder** Docker image to trigger a pipeline in Azure DevOps (ADO). It passes parameters using REST API provided by ADO. The definition of parameters passed through REST API is taken from kaniko-build-config located on the `main` branch of the `test-infra` repository and GitHub context.
```

> During the ADO pipeline execution, the image builder action is checks for the status to be reported. When the execution ends, it fetches the status and logs.

Reviewer 4: ```suggestion
During the ADO pipeline execution, the image-builder action checks for the status to be reported. When the execution ends, it fetches the status and logs.
```

> If the execution failed, the the image builder action fails. If the execution succeeds, it sets the output and success status.

Reviewer 4: ```suggestion
If the execution fails, the image-builder action fails. If the execution succeeds, it sets the output and success status.
```
